.gitlab-ci.yml

```diff
@@ -1,259 +1,361 @@
-    - gem list -ie "${GEM_BUNDLE_VER[@]}" --silent bundler || "${GEM_INSTALL[@]}" --local "${GEM_BUNDLE_VER[@]}" bundler || "${GEM_INSTALL[@]}" "${GEM_BUNDLE_VER[@]}" bundler
-    - 'rm -rf pkg/ || :'
-    - bundle check || rm -f Gemfile.lock && ("${BUNDLER_INSTALL[@]}" --local || "${BUNDLER_INSTALL[@]}")
-
-
-.validation_checks: &validation_checks
-  script:
-    - bundle exec rake syntax
-    - bundle exec rake check:dot_underscore
-    - bundle exec rake check:test_file
-    - bundle exec rake pkg:check_version
-    - bundle exec rake pkg:compare_latest_tag
-    - bundle exec rake lint
-    - bundle exec rake clean
-    - bundle exec puppet module build
-
-.spec_tests: &spec_tests
+      - 'Gemfile.lock'
+
+.unit_base: &unit_base
+  stage: 'unit'
+  tags: ['docker']
+  variables:
+    STRINGIFY_FACTS: 'no'
+    TRUSTED_NODE_DATA: 'yes'
   script:
-    - bundle exec rake spec
+    - 'bundle exec rake spec'
+  cache:
+    policy: 'pull'
+  dependencies: []
+  artifacts:
+    when: 'always'
+    paths:
+      - 'Gemfile.lock'
 
-stages:
-  - validation
-  - unit
-  - acceptance
-  - deploy
-
-# Puppet 4.7 for PE 2016.4 LTS Support (EOL: 2018-10-21)
-# See: https://puppet.com/misc/puppet-enterprise-lifecycle
-# --------------------------------------
-pup4_7-validation:
-  stage: validation
-  tags:
-    - docker
-  image: ruby:2.1
+.acceptance_base: &acceptance_base
+  stage: 'acceptance'
+  tags: ['beaker']
   variables:
-    PUPPET_VERSION: '~> 4.7.0'
-  <<: *cache_bundler
-  <<: *setup_bundler_env
-  <<: *validation_checks
+    BUNDLE_WITHOUT: 'development'
+  cache:
+    policy: 'pull'
+  dependencies: []
+  artifacts:
+    when: 'always'
+    paths:
+      - 'Gemfile.lock'
 
-pup4_7-unit:
-  stage: unit
-  tags:
-    - docker
-  image: ruby:2.1
+# ----------------------------------------------------------------------
+# Version Matrix
+#
+# Selections are based on the component version combinations included with
+# currently supported releases of SIMP and Puppet Enterprise.
+#
+# For more information see:
+#  * https://puppet.com/docs/pe/latest/overview/component_versions_in_recent_pe_releases.html
+#  * https://puppet.com/misc/puppet-enterprise-lifecycle
+#
+# | Release   | Puppet | Ruby | End-of-Life Date |
+# |-----------|--------|------|------------------|
+# | PE 2016.4 | 4.7*   | 2.1  | 2018-10
+# | PE 2016.5 | 4.8    | 2.1  | 2017-05
+# | SIMP 6.0  | 4.8    | 2.1  | TBD
+# | PE 2017.1 | 4.9    | 2.1  | 2017-10
+# | PE 2017.2 | 4.10   | 2.1  | 2018-02
+# | SIMP 6.1  | 4.10   | 2.1  | TBD
+# | PE 2017.3 | 5.3    | 2.4  | 2018-08
+# | PE 2018.1 | 5.5    | 2.4  | 2020-05
+#
+# > *PE 2016.4 released with Puppet 4.7.0, but upgraded to the 4.10 series
+# > starting with 2016.4.5.
+#
+
+.pe_2016_4: &pe_2016_4
   variables:
     PUPPET_VERSION: '~> 4.7.0'
-  <<: *cache_bundler
-  <<: *setup_bundler_env
-  <<: *spec_tests
 
+.simp_6_0: &simp_6_0
+  variables:
+    PUPPET_VERSION: '~> 4.8.1'
```

Comment:
These `$PUPPET_VARIABLE` declarations were much better off when they were defined in the jobs.   It was slightly redundant, but much clearer to read and maintain.  

Abstracting them into YAML merge keys takes twice as many lines to do the same thing, and is significantly more oblique.

The merge keys are misleadingly named after product releases (e.g., `simp_6_0`, `pe_2016_4`):
  * Their only function is to provide the `$PUPPET_VERSION` variable,
  * All jobs that use these keys are based on the Puppet version (e.g., `el-pup_4_8-default`, `el-pup_4_8-fips`)
  * Multiple SIMP / PE products can require the same puppet version.